src/main/java/com/faforever/client/login/LoginController.java

```diff
@@ -396,6 +297,49 @@ public void onDownloadUpdateButtonClicked() {
         });
   }
 
+  public void onLoginButtonClicked() {
+    initializeFuture.join();
+
+    clientProperties.getServer()
+        .setHost(serverHostField.getText())
+        .setPort(Integer.parseInt(serverPortField.getText()));
+
+    clientProperties.getReplay()
+        .setRemoteHost(replayServerHostField.getText())
+        .setRemotePort(Integer.parseInt(replayServerPortField.getText()));
+
+    clientProperties.getIrc()
+        .setHost(ircServerHostField.getText())
+        .setPort(Integer.parseInt(ircServerPortField.getText()));
+
+    clientProperties.getApi().setBaseUrl(apiBaseUrlField.getText());
+    clientProperties.getOauth().setBaseUrl(oauthBaseUrlField.getText());
+
+    OAuthValuesReceiver oAuthValuesReceiver = applicationContext.getBean(OAuthValuesReceiver.class);
```

Comment:
Why retrieve this via the application context rather than the constructor?

Reply:
Because it's not a singleton but a prototype bean:
```java
@Component
@Scope(ConfigurableBeanFactory.SCOPE_PROTOTYPE)
@RequiredArgsConstructor
@Slf4j
public class OAuthValuesReceiver {
```
which is needed because it's stateful.

If it was injected via constructor, LoginController would receive only once instance and in order for it to be able to re-use this instance, the value receiver would need to be mutable (resetting `addressFuture`). Immutability, however, makes it easier to write, understand and use.